cranelift-codegen/src/simple_preopt.rs

```diff
@@ -536,11 +542,245 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
     }
 }
 
+struct BranchOptInfo {
+    br_inst: Inst,
+    cmp_arg: Value,
+    args: ValueList,
+    new_opcode: Opcode,
+}
+
+/// Fold comparisons into branch operations when possible.
+///
+/// This matches against operations which compare against zero, then use the
+/// result in a `brz` or `brnz` branch. It folds those two operations into a
+/// single `brz` or `brnz`.
+fn branch_opt(pos: &mut FuncCursor, inst: Inst) {
+    let mut info = if let InstructionData::Branch {
+        opcode, ref args, ..
+    } = pos.func.dfg[inst]
+    {
+        let first_arg = {
+            let args = pos.func.dfg.inst_args(inst);
+            args[0]
+        };
+
+        let iconst_inst =
+            if let ValueDef::Result(iconst_inst, _) = pos.func.dfg.value_def(first_arg) {
+                iconst_inst
+            } else {
+                return;
+            };
+
+        if let InstructionData::IntCompareImm {
+            opcode: Opcode::IcmpImm,
+            arg: cmp_arg,
+            cond: cmp_cond,
+            imm: cmp_imm,
+        } = pos.func.dfg[iconst_inst]
+        {
+            let cmp_imm: i64 = cmp_imm.into();
+            if cmp_imm != 0 {
```

Comment:
Doing this would require changing the definition of `Imm64(i64)` to `Imm64(pub i64)`, as far as I can tell. I could certainly do that, but figured there might be a reason why it's private.

I'd love to do that as `Imm64::from(0)` but https://github.com/rust-lang/rust/issues/57240 is not yet a thing.